lib/SIL/IR/OperandOwnership.cpp

```diff
@@ -127,559 +122,368 @@ SHOULD_NEVER_VISIT_INST(GetAsyncContinuation)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, StructElementAddr)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, SwitchEnumAddr)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, SwitchValue)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, TailAddr)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, ThickToObjCMetatype)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, ThinFunctionToPointer)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, ThinToThickFunction)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, TupleElementAddr)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, UncheckedAddrCast)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, UncheckedRefCastAddr)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, UncheckedTakeEnumDataAddr)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, UnconditionalCheckedCastAddr)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, AllocValueBuffer)
-CONSTANT_OWNERSHIP_INST(None, NonLifetimeEnding, DeallocValueBuffer)
+// Instructions that currently violate structural ownership requirements,
+// and therefore completely defeat canonicalization and optimization of any
+// OSSA value that they use.
+OPERAND_OWNERSHIP(PointerEscape, ProjectBox) // The result is a T*.
+OPERAND_OWNERSHIP(PointerEscape, ProjectExistentialBox)
+OPERAND_OWNERSHIP(PointerEscape, UncheckedOwnershipConversion)
+
+// Instructions that escape reference bits with unenforced lifetime.
+OPERAND_OWNERSHIP(BitwiseEscape, UncheckedBitwiseCast)
+OPERAND_OWNERSHIP(BitwiseEscape, ValueToBridgeObject)
+OPERAND_OWNERSHIP(BitwiseEscape, RefToRawPointer)
+OPERAND_OWNERSHIP(BitwiseEscape, UncheckedTrivialBitCast)
+// FIXME: verify that no-escape results are always trivial
+OPERAND_OWNERSHIP(BitwiseEscape, ConvertEscapeToNoEscape)
+
+// Instructions that end the lifetime of an owned value.
+OPERAND_OWNERSHIP(DestroyingConsume, AutoreleaseValue)
+OPERAND_OWNERSHIP(DestroyingConsume, DeallocBox)
+OPERAND_OWNERSHIP(DestroyingConsume, DeallocExistentialBox)
+OPERAND_OWNERSHIP(DestroyingConsume, DeallocRef)
+OPERAND_OWNERSHIP(DestroyingConsume, DestroyValue)
+OPERAND_OWNERSHIP(DestroyingConsume, EndLifetime)
+OPERAND_OWNERSHIP(DestroyingConsume, BeginCOWMutation)
+OPERAND_OWNERSHIP(DestroyingConsume, EndCOWMutation)
+OPERAND_OWNERSHIP(ForwardingConsume, MarkUninitialized)
+
+// Instructions that move an owned value.
+OPERAND_OWNERSHIP(ForwardingConsume, CheckedCastValueBranch)
+OPERAND_OWNERSHIP(ForwardingConsume, UnconditionalCheckedCastValue)
+OPERAND_OWNERSHIP(ForwardingConsume, InitExistentialValue)
+OPERAND_OWNERSHIP(ForwardingConsume, DeinitExistentialValue)
+OPERAND_OWNERSHIP(ForwardingConsume, Throw)
+
+// Instructions that expose a pointer within a borrow scope.
+OPERAND_OWNERSHIP(InteriorPointer, RefElementAddr)
+OPERAND_OWNERSHIP(InteriorPointer, RefTailAddr)
+OPERAND_OWNERSHIP(InteriorPointer, OpenExistentialBox)
+// FIXME: OpenExistential[Box]Value should be a forwarding consume.
+OPERAND_OWNERSHIP(InteriorPointer, OpenExistentialValue)
```

Comment:
I think that open_existential_box{_value} are forwarded borrow instructions, no?

Reply:
yes, the `open_existential_box` instructions should be `ForwardingBorrow`. I don't understand why they can't take owned values, but I'll worry about that later.